### DISCUSSION TITLE
Keychron keyboards firmware upgrade support

### DISCUSSION THREAD

**Discussion by marcosps** [fwupd/fwupd#5664] [General]
**Describe the question**
Do you have plans to add support for keychron keyboards? I'm asking because I opened a bug on HID kernel modules about a warning that I'm seeing, and it's possible that it's related to the keyboard firmware.

Either way, I sent a message to Keychron contact web page about fwupd support, but no answer yet.

Thanks in advance!

**Comment by hughsie** [comment]
There's no plan right now, but there are a few people interested in implementing this. Unless we can get buy-in from Keychron to upload firmware to the LVFS I don't think there's a lot of point adding a plugin IMHO.

**Reply by marcosps** [reply]
I received an answer to the contact email saying the request will be passed to the technical team for consideration. @hughsie based on your experience, does it mean they won't be doing it at all? I don't know any other communication channel with them...

**Reply by hughsie** [reply]
I think it's worth emailing them with links to the lvfs -- and feel free to cc me -- richard_at_hughsie.com -- i'll happily add my introduction. Thanks.

**Reply by marcosps** [reply]
Done, you're now added to my last email to them, let's see how it goes :)

Thanks a lot!

**Reply by devurandom** [reply]
Flashing of my Keychron K9 Pro is done through the QMK CLI (https://docs.qmk.fm/#/newbs_flashing?id=flash-your-keyboard-from-the-command-line), which (judging from the console output) seems to utilize dfu-util (https://dfu-util.sourceforge.net/). System76 advertises their Launch keyboard (https://system76.com/accessories/launch) as also using QMK and providing firmware updates via LVFS. Maybe another fwupd plugin is not even needed, but Keychron only needs to permit LVFS to distribute the firmware files?

**Reply by hughsie** [reply]
Can you grab the full dfu-util command line from anywhere? It might *just work* :)

**Reply by marcosps** [reply]
Would that work even without QMK support? My keyboard is a simple K8 :)

**Reply by devurandom** [reply]
> Would that work even without QMK support?

Probably not. :( 

> Can you grab the full dfu-util command line from anywhere? It might just work :)

I found the code responsible for the `qmk flash` command: https://github.com/qmk/qmk_firmware/blob/4e369d405af6bba1adce6337b2e1b1ea1788566c/lib/python/qmk/flashers.py
(Called from https://github.com/qmk/qmk_firmware/blob/4e369d405af6bba1adce6337b2e1b1ea1788566c/lib/python/qmk/cli/flash.py.)

You can find the firmware file of my keyboard behind the "Download K9 Pro ANSI RGB Version Firmware" button on https://www.keychron.com/blogs/archived/k9-pro-factory-reset-and-firmware-flash.
(Other keyboards' firmware is linked on https://www.keychron.com/pages/firmware, but not all of them use QMK.)
The source code for my keyboard is supposed to be https://github.com/Keychron/qmk_firmware/tree/bluetooth_playground/keyboards/keychron/k8_pro, but I did not try to build let alone flash it.

**Reply by hughsie** [reply]
Super useful. We support all those protocols already (uf2, dfu, dfuse, st-dfu, etc) with one exception: SAM-BA (which is what mdloader provides). I think the hard bit is to convince Keychron to test with fwupd, and upload firmware to the LVFS.

**Reply by devurandom** [reply]
I sent them an email with you, @hughsie, in CC. Please feel free to correct or amend anything I wrote in that email.
